File: lib/wicket-utils/src/main/java/br/net/mirante/singular/util/wicket/behavior/DatePickerInitBehaviour.java
/*
 * Copyright (c) 2016, Mirante and/or its affiliates. All rights reserved.
 * Mirante PROPRIETARY/CONFIDENTIAL. Use is subject to license terms.
 */

package br.net.mirante.singular.util.wicket.behavior;

import org.apache.wicket.Component;
import org.apache.wicket.MarkupContainer;
import org.apache.wicket.markup.html.form.FormComponent;
import org.apache.wicket.util.visit.IVisit;
import org.apache.wicket.util.visit.IVisitor;

import br.net.mirante.singular.util.wicket.bootstrap.datepicker.BSDatepickerConstants;

public class DatePickerInitBehaviour extends InitScriptBehaviour {

    @Override
    public String getScript(Component component) {

        String js = "";

        js += " $('#%s').datepicker({ ";
        js += "     rtl: App.isRTL(), ";
        js += "     orientation: 'right', ";
        js += "     autoclose: true, ";
        js += "     language: 'pt-BR' ";
        js += " }) ";

        String idDatepicker = component.getMarkupId();
        String idInput = component.getMarkupId();

        if (component instanceof MarkupContainer) {
            FormComponent fc = ((MarkupContainer) component).visitChildren(FormComponent.class, new IVisitor<FormComponent, FormComponent>() {
                @Override
                public void component(FormComponent object, IVisit<FormComponent> visit) {
                    visit.stop(object);
                }
            });
            if (fc != null) {
                idInput = fc.getMarkupId();
            }
        }

        js += ".on('changeDate', function(){";
        js += String.format("var input = $('#%s');", idInput);
        js += String.format("var format = $('#%s').attr('data-date-format').toUpperCase();", idDatepicker);
        js += "     if( format == 'DD/MM/YYYY' && /\\d{1,2}\\/\\d{1,2}\\/\\d{4}/.test(input.val()) ";
        js += "         || format == 'DD/MM' && /\\d{1,2}\\/\\d{1,2}/.test(input.val())) { ";
        js += "         input.trigger('" + BSDatepickerConstants.JS_CHANGE_EVENT + "');";
        js += "     } ";
        js += " }); ";

        return String.format(js, idDatepicker);
    }
}

File: lib/wicket-utils/src/main/java/org/opensingular/lib/wicket/util/behavior/DatePickerInitBehaviour.js
(function () {
    var $datepicker = $("#${datePickerMarkupId}");
    var $input = $("#${inputMarkupId}");

    var settings = {
        rtl: App.isRTL(),
        orientation: 'right',
        autoclose: true,
        language: 'pt-BR'
    };

    if (${configureBeforeShowDay}) {
        settings.beforeShowDay = function (date) {
            var dd = date.getDate();
            if (dd.toString().length === 1) {
                dd = "0" + dd;
            }
            var mm = date.getMonth() + 1;
            if (mm.toString().length === 1) {
                mm = "0" + mm;
            }
            var yyyy = date.getFullYear();
            return ${enabledDates}.indexOf(dd + "/" + mm + "/" + yyyy) >= 0;
        }
    }

    if (!$input.prop('disabled')) {
        $datepicker.datepicker(settings).on('changeDate', function () {
            var input = $input;
            var format = $datepicker.data('dateFormat').toUpperCase();
            if (format === 'DD/MM/YYYY' && /\d{1,2}\/\d{1,2}\/\d{4}/.test(input.val())
                || format === 'DD/MM' && /\d{1,2}\/\d{1,2}/.test(input.val())) {
                input.trigger('${changeEvent}');
            }
        });
    }
})();